### PR TITLE
Add support for two finger scroll and right click with trackpad

### LINKED ISSUE #1383: [Issue]: Physical trackpad input is not fully functioning
### Describe the bug

When connected to a physical trackpad, only moving the cursor is supported, tap as click/two finger scroll/two finger tap as right click all doesn't work. Physical click button works.

### Steps to reproduce

1. Connect the trackpad to your android device
2. Try use the trackpad as you'll normally do on a laptop

### Affected games

ANY

### Other Moonlight clients

PC

### Moonlight adjusted settings

No

### Moonlight adjusted settings (please complete the following information)

none

### Moonlight default settings

Yes

### Gamepad-related connection issue

No

### Gamepad-related input issue

Yes

### Gamepad-related streaming issue

No

### Android version

Android 14

### Device model

OnePlus Pad Pro

### Server PC OS version

Windows11 23H2

### Server PC GeForce Experience version

-

### Server PC Nvidia GPU driver version

-

### Server PC antivirus and firewall software

-

### Screenshots

_No response_

### Relevant log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ClassicOldSong**
Resolves https://github.com/moonlight-stream/moonlight-android/issues/1383

### CODE REVIEW

**Comment by ClassicOldSong** [comment]
Download build at https://github.com/ClassicOldSong/moonlight-android/releases/tag/v12.1-BLACK

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]
I think some of the if/else nesting can be improved.

File: app/src/main/java/com/limelight/binding/input/touch/TrackpadContext.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                            } else {
+                                conn.sendMouseHighResScroll((short)(deltaY * SCROLL_SPEED_FACTOR_Y));
+                                if (absDeltaX * 1.05 >= absDeltaY) {
+                                    conn.sendMouseHighResHScroll((short)(-deltaX * SCROLL_SPEED_FACTOR_X));
+                                }
+                            }
+                        }
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (ClassicOldSong) [reply app/src/main/java/com/limelight/binding/input/touch/TrackpadContext.java:193]: The multi-finger implementation originally written is pertty hard to maintain. It has multiple states that can only be handled like this.

It's better if the multi-finger handler could be refactored completely.

File: app/src/main/java/com/limelight/binding/input/touch/TrackpadContext.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        } else {
+            // Second finger released, should trigger right click immediately
+            if (pointerCount == 1 && !confirmedMove) {
+                conn.sendMouseButtonDown(MouseButtonPacket.BUTTON_RIGHT);
+                conn.sendMouseButtonUp(MouseButtonPacket.BUTTON_RIGHT);
+                isClickPending = false;
+                isDblClickPending = false;
+                confirmedDrag = false;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (ClassicOldSong) [reply app/src/main/java/com/limelight/binding/input/touch/TrackpadContext.java:98]: Code style and syntax goes last.

File: app/src/main/java/com/limelight/Game.java [hunk 3]
```diff
@@ -1972,145 +1998,153 @@ else if (event.getActionMasked() == MotionEvent.ACTION_UP || event.getActionMask
...
+            else {
+                if (eventSource == InputDevice.SOURCE_TOUCHPAD) {
+                    return handleTouchInput(event, 0.f, 0.f, trackpadContextMap);
+                } else {
+                    if (virtualController != null &&
+                            (virtualController.getControllerMode() == VirtualController.ControllerMode.MoveButtons ||
+                                    virtualController.getControllerMode() == VirtualController.ControllerMode.ResizeButtons)) {
+                        // Ignore presses when the virtual controller is being configured
+                        return true;
+                    }
```
[the remark this answers is not in the document]
Comment (ClassicOldSong) [reply app/src/main/java/com/limelight/Game.java:2010]: For clearity issue, I prefer nested `if` in the `else`.

File: app/src/main/java/com/limelight/binding/input/touch/TrackpadContext.java [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        long timeDelta = eventTime - originalTouchTime;
+        return isWithinTapBounds(lastTouchX, lastTouchY) && timeDelta <= TAP_TIME_THRESHOLD;
+    }
+
+    private byte getMouseButtonIndex() {
+        if (pointerCount == 2) {
+            return MouseButtonPacket.BUTTON_RIGHT;
+        } else {
+            return MouseButtonPacket.BUTTON_LEFT;
+        }
```
[the remark this answers is not in the document]
Comment (ClassicOldSong) [reply app/src/main/java/com/limelight/binding/input/touch/TrackpadContext.java:71]: https://github.com/moonlight-stream/moonlight-android/blob/master/app/src/main/java/com/limelight/binding/input/touch/RelativeTouchContext.java#L138-L146

**Comment by ClassicOldSong** [comment]
Closing this PR since it's actually helping people out.

What a drama in the discord server!

Getting kicked for helping people out when the main project has no updates and no message about updates for months, get slammed by "you fixed the issue you opened doesn't necessarily mean others have the same issue", "you fixed your problem but might break every else" and "Either be patient or go away".

Sure, I'll go away, no more PRs to help people actually in need and I still wish good luck to you.

**Pull request closed without merging**